File: tools/PCLServer/PclRpcClient.cpp
#include <stdio.h>
#include <iostream>
#include <getopt.h>
#include <sstream>

#include <thrift/protocol/TBinaryProtocol.h>
#include <thrift/transport/TSocket.h>
#include <thrift/transport/TTransportUtils.h>

#include "seekerror.h"
#include "gen-cpp/PclRPC.h"
#include "gen-cpp/pcl_rpc_constants.h"

using namespace std;
using namespace PclRPC;
using namespace apache::thrift;
using namespace apache::thrift::protocol;
using namespace apache::thrift::transport;

struct Args
{
    string species;
    vector<string> genes;
    vector<string> datasets;
    uint32_t port = 9010;
    bool verbose = false;
};

bool parseArgs(int argc, char **argv, Args &args)
{
    string usage = "Usage: PclRpcClient -s <species> -g <comma-sep-query-genes> -d <comma-sep-datasets> \n"
                   "Run a PCL query for the given species, genes and datasets.\n";

    static struct option long_options[] = {
        {"species", required_argument, 0, 's'},
        {"genes", required_argument, 0, 'g'},
        {"datasets", required_argument, 0, 'd'},
        {"port", required_argument, 0, 'p'},
        {"verbose", no_argument, 0, 'v'},
        {0, 0, 0, 0}};
    int opt = 0;
    int long_index = 0;
    while ((opt = getopt_long(argc, argv, "s:g:d:p:v",
                              long_options, &long_index)) != -1)
    {
        switch (opt)
        {
        case 's':
            args.species = optarg;
            break;
        case 'g':
        {
            stringstream s_stream(optarg);
            while (s_stream.good())
            {
                string gene;
                getline(s_stream, gene, ',');
                args.genes.push_back(gene);
            }
        }
        break;
        case 'd':
        {
            stringstream s_stream(optarg);
            while (s_stream.good())
            {
                string dataset;
                getline(s_stream, dataset, ',');
                args.datasets.push_back(dataset);
            }
        }
        break;
        case 'p':
            args.port = atoi(optarg);
            break;
        case 'v':
            args.verbose = true;
            break;
        default:
            cerr << "Error: unrecognized options: " << opt << endl;
            cerr << usage << endl;
            return false;
        }
    }
    if (args.genes.size() == 0 || args.datasets.size() == 0)
    {
        cerr << "ERROR: Must specify both query genes and datasets" << endl;
        cerr << usage << endl;
        return false;
    }

    return true;
}

int main(int argc, char **argv)
{
    Args args;
    bool res = parseArgs(argc, argv, args);
    if (res == false)
    {
        exit(-1);
    }

    if (args.verbose == true) {
        std::cout << args.species << endl;
        for (auto gene : args.genes) {
            std::cout << gene << ",";
        }
        std::cout << endl;
    }

    shared_ptr<TTransport> socket(new TSocket("localhost", args.port));
    shared_ptr<TTransport> transport(new TBufferedTransport(socket));
    shared_ptr<TProtocol> protocol(new TBinaryProtocol(transport));
    PclRPCClient pclClient(protocol);

    try
    {
        transport->open();
        pclClient.ping();

        // // Check for version compatibility
        int32_t version = pclClient.getRpcVersion();
        assert(version == g_pcl_rpc_constants.PclRPCVersion);

        PclQueryArgs pclQueryArgs;
        PclResult result;
        pclQueryArgs.species = args.species;
        pclQueryArgs.__isset.settings = true;
        pclQueryArgs.settings.__set_outputGeneExpression(true);
        pclQueryArgs.settings.__set_outputNormalized(true);
        pclQueryArgs.settings.__set_rbp(-1);
        // query.settings.__set_outputGeneCoexpression(true);
        // query.settings.__set_outputQueryExpression(true);
        // query.settings.__set_outputQueryCoexpression(true);
        pclQueryArgs.__set_genes(args.genes);
        pclQueryArgs.__set_datasets(args.datasets);

        pclClient.pclQuery(result, pclQueryArgs);
        if (result.success == false) {
            throw query_error(result.statusMsg);
        }
        assert(result.geneExpressions.size() > 0);

        int numDatasets = args.datasets.size();
        int numGenes = args.genes.size();
        int offset = 0;
        for (int i=0; i<numDatasets; i++) {
            for (int j=0; j<numGenes; j++) {
                cout << "dset: " << i << ", gene: " << j << endl;
                int numSamples = result.datasetSizes[i];
                for(int k=0; k<numSamples; k++) {
                    // Can't use offset calculation because each dataseSize is different
                    // int offset = k + j*datasetSize + i*datasetSize*numGenes; 
                    cout << "val: " << result.geneExpressions[offset] << endl;
                    offset++;
                }
            }
        }

    } catch (TException &tx) {
        std::cout << "ERROR: " << tx.what() << endl;
        exit(-1);
    } catch (exception &err) {
        std::cout << "ERROR: " << err.what() << endl;
        exit(-1);
    }

    transport->close();
}

File: tools/PCLServer/PclRpcClient.py
import time
import argparse
import importlib
from thrift.transport import TTransport, TSocket
from thrift.protocol.TBinaryProtocol import TBinaryProtocol
# Since the path gen-py/seek_rpc has a hyphen in it we need to use
#   the import_module() function rather than the usual import call
PclRPC = importlib.import_module('gen-py.pcl_rpc.PclRPC')
ttypes = importlib.import_module('gen-py.pcl_rpc.ttypes')
constants = importlib.import_module('gen-py.pcl_rpc.constants')

host = 'localhost'

'''
Query parameters and options:
struct PclSettings {
    1: optional bool outputNormalized = false;
    2: optional bool outputGeneExpression = false;
    3: optional bool outputQueryExpression = false;
    4: optional bool outputGeneCoexpression = false;
    5: optional bool outputQueryCoexpression = false;
    6: optional double rbp = 0.99;
}

struct PclQueryArgs {
    1: required string species = "Unknown";
    2: required list<string> genes;
    3: required list<string> datasets;
    4: optional list<string> query;
    5: optional PclSettings settings;
    6: optional string outputDir = "/tmp/seek";
}

struct PclResult {
    1: required bool success;
    2: required list<i32> datasetSizes;
    3: required list<double> geneExpressions;
    4: optional list<double> geneCoexpressions;
    5: optional list<double> queryExpressions;
    6: optional list<double> queryCoexpressions;
    7: optional PclStatus status;
    8: optional string statusMsg;
}
'''

def runQuery(args):
    socket = TSocket.TSocket(host, args.port)
    transport = TTransport.TBufferedTransport(socket)
    protocol = TBinaryProtocol(transport)
    client = PclRPC.Client(protocol)
    transport.open()
    version = client.getRpcVersion()
    assert version == constants.PclRPCVersion

    settings = PclRPC.PclSettings(
        outputNormalized = True,
        outputGeneExpression = True,
        outputQueryExpression = False,
        outputGeneCoexpression = False,
        outputQueryCoexpression = False,
        rbp = -1,
    )

    pclArgs = PclRPC.PclQueryArgs(species=args.species, genes=args.genes, datasets=args.datasets, settings=settings)

    retval = -1
    result = client.pclQuery(pclArgs)
    if result.success is True:
        # Output to file if requested
        if args.outfile is not None:
            with open(args.outfile, "w") as fp:
                fp.write("\n".join(str(round(item, 6)) for item in result.geneExpressions))
        # Print to command line
        numGenes = len(args.genes)
        offset = 0;
        for i, numSamples in enumerate(result.datasetSizes):
            for j in range(numGenes):
                print(f'dset: {i}, gene: {j}')
                for k in range(numSamples):
                    print(f'val: {result.geneExpressions[offset]}')
                    offset += 1
        retval = 0
    else:
        print(f'query error: {result.statusMsg}')
        retval = -1

    transport.close()
    return retval


if __name__ == "__main__":
    argParser = argparse.ArgumentParser()
    argParser.add_argument('--species', '-s', default='human', type=str,
                           help='species name')
    argParser.add_argument('--genes', '-g', default=None, type=str,
                           help='list of genes to query')
    argParser.add_argument('--datasets', '-d', default=None, type=str,
                           help='list of datasets to query')
    argParser.add_argument('--outfile', '-o', default=None, type=str,
                           help='filename to output results')
    argParser.add_argument('--port', '-p', default=9010, type=int,
                           help='server port')
    args = argParser.parse_args()

    if args.genes is None or args.datasets is None:
        print('Please specify both genes and datasets, use [-g gene1,gene2,...] [-d dset1,dset2,...]')
        exit(-1)
    args.genes = args.genes.split(',')
    args.datasets = args.datasets.split(',')

    retval = runQuery(args)
    exit(retval)


#species = 'human'
#genes = ['SMO', 'PTCH1', 'PTCH2', 'BOC'] - this one differs from web, ambiguous symbol->entrez?
#genes = ['90634', '23659']

#species = 'fly'
#genes = ['35234', '35232']
#genes = ['34930', '35234', '35232']

#species = 'yeast'
#genes = ['YGL142C', 'YHR188C']

# species = 'mock'
# genes = ['six', 'four']
# genes = ['90634', '23659']
